dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/clients/common/UserInfoHelper.java

```diff
@@ -0,0 +1,130 @@
+    }
+
+    /**
+     * get userinfo from provider's UserInfo Endpoint if configured and active.
+     * If successful, update properties in the ProviderAuthenticationResult
+     *
+     * @return true if PAR was updated with userInfo
+     *
+     */
+    public boolean getUserInfo(ProviderAuthenticationResult oidcResult,
+            SSLSocketFactory sslSocketFactory, String accessToken, String subjectFromIdToken) {
+
+        if (!willRetrieveUserInfo() || accessToken == null) {
+            return false;
+        }
+        String userInfoStr = getUserInfoFromURL(clientConfig, sslSocketFactory, accessToken);
+
+        if (userInfoStr == null) {
+            return false;
+        }
+
+        if (!isUserInfoValid(userInfoStr, subjectFromIdToken)) {
+            return false;
+        }
+
+        updateAuthenticationResultPropertiesWithUserInfo(oidcResult, userInfoStr);
+        return true;
+    }
+
+    protected void updateAuthenticationResultPropertiesWithUserInfo(ProviderAuthenticationResult oidcResult, String userInfoStr) {
+        oidcResult.getCustomProperties().put(Constants.USERINFO_STR, userInfoStr);
+    }
+
+    // per oidc-connect-core-1.0 sec 5.3.2, sub claim of userinfo response must match sub claim in id token.
+    protected boolean isUserInfoValid(String userInfoStr, String subClaim) {
+        String userInfoSubClaim = getUserInfoSubClaim(userInfoStr);
+        if (userInfoSubClaim == null || subClaim == null || userInfoSubClaim.compareTo(subClaim) != 0) {
+            Tr.error(tc, "USERINFO_INVALID", new Object[] { userInfoStr, subClaim });
+            return false;
+        }
+        return true;
+    }
+
+    protected String getUserInfoSubClaim(String userInfo) {
+        JSONObject jobj = null;
+        try {
+            jobj = JSONObject.parse(userInfo);
+        } catch (Exception e) { // ffdc
+        }
+        return jobj == null ? null : (String) jobj.get("sub");
+    }
+
+    /**
+     * Obtain userInfo from an OIDC provider
+     *
+     * @return the userInfo string, or null
+     */
+    protected String getUserInfoFromURL(ConvergedClientConfig config, SSLSocketFactory sslsf, String accessToken) {
+        String url = config.getUserInfoEndpointUrl();
+        boolean hostnameVerification = config.isHostNameVerificationEnabled();
+
+        // https required by spec, use of http is not spec compliant
+        if (!url.toLowerCase().startsWith("https:") && config.isHttpsRequired()) {
+            Tr.error(tc, "OIDC_CLIENT_URL_PROTOCOL_NOT_HTTPS", new Object[] { url });
+            return null;
+        }
+
+        OidcClientUtil oidccu = new OidcClientUtil();
+        int statusCode = 0;
+        String responseStr = null;
+        try {
+            Map<String, Object> resultMap = oidccu.getUserinfo(url, accessToken, sslsf, hostnameVerification);
+            if (resultMap == null) {
+                throw new Exception("result map from getUserinfo is null");
```

Comment:
This needs to be an NLS message. This relies on our FFDC catching/logging framework to automatically generate an FFDC for an error condition using a string that won't be translated. That goes against our serviceability guidelines.

Reply:
it should never happen so I didn't think it worthy of nls but did put some text in there just in case I was wrong.